Allow user-defined stream implementations with read/seek/size/close callbacks

tmm1 opened the pull request:
Added as a new client sub-api, ala opengl_cb.

Once the sub-api is initialized with callbacks, you can create a stream pointed at `cb://12345` where `12345` is a unique cookie that identifies the stream. This cookie is passed as the first argument to the registered callbacks.

wm4 commented:
- the callbacks need somehow be passed down to stream_cb.c, instead of using global mutable state
- off_t and ssize_t cause more trouble than they could ever solve; using int64_t (and maybe ptrdiff_t) is preferable, since they exist in C90/C99/C++ on all platforms
- turning strings into numbers and casting the to pointers is weird (if we do this, we might as well encode all callbacks this way on the URI?) - maybe it would be better to have a separate open callback, which would return an opaque user pointer for a given URI string
- this needs to document that the stream callbacks can not use most of the libmpv API, since this would lead to deadlocks
- possibly a mechanism to interrupt blocking operation is needed (stream_file.c for example uses poll() on the file handle and a notification pipe, which enables mpv to exit even when reading from a blocked pipe)

tmm1 commented:
> the callbacks need somehow be passed down to stream_cb.c, instead of using global mutable state

At a loss on how to do this best. stream_t doesn't allow access to mpv_handle or MPContext. I tried putting the callback struct in MPOpts, but it wasn't copied over to the stream_t opts.

> off_t and ssize_t cause more trouble than they could ever solve; using int64_t (and maybe ptrdiff_t) is preferable, since they exist in C90/C99/C++ on all platforms

Switched to int64_t

> this needs to document that the stream callbacks can not use most of the libmpv API, since this would lead to deadlocks

Done

> turning strings into numbers and casting the to pointers is weird (if we do this, we might as well encode all callbacks this way on the URI?) - maybe it would be better to have a separate open callback, which would return an opaque user pointer for a given URI string

I tried to cram all the callbacks into the string initially, but it got messy quickly. A custom `open` is a good idea, I've switched to that here.

> possibly a mechanism to interrupt blocking operation is needed (stream_file.c for example uses poll() on the file handle and a notification pipe, which enables mpv to exit even when reading from a blocked pipe)

This could certainly be useful, although I'm not sure how best to design the api

wm4 commented:
> At a loss on how to do this best. stream_t doesn't allow access to mpv_handle or MPContext. I tried putting the callback struct in MPOpts, but it wasn't copied over to the stream_t opts.

It's not necessarily simple - it's not clear how this should be passed through at all. One could argue that this is a good case to add to ``mpv_global``, though I'm not sure if I love that idea.

In any case, ``mpv_stream_cb_set_callbacks()`` should be made thread-safe too.

``MPOpts`` (and ``mpv_global`` as well) is copied for the demuxer thread to avoid thread-safety issues. (The code is in misc.c ``create_sub_global()``.)

>>    possibly a mechanism to interrupt blocking operation is needed (stream_file.c for example uses poll() on the file handle and a notification pipe, which enables mpv to exit even when reading from a blocked pipe)

> This could certainly be useful, although I'm not sure how best to design the api

Well, maybe this can be left for later. Seems pretty orthogonal to add. But ``mpv_stream_cb_set_callbacks()`` could be annoying here (when this gets implemented later, a new function would have to be added to set whatever new callbacks are needed for it).

tmm1 commented:
> Also, I think close() shouldn't be allowed to fail, if that's the intention of the int return value. POSIX close() can in theory fail, but that's misdesigned.

Done.

> I think open() should get an opaque user pointer as parameter, which is set with mpv_stream_cb_set_callbacks().

Done. I added a `mpv_stream_cb_init()` to set the user_data pointer.

> Should we account for the possibility that multiple mpv_handles want to set different stream callbacks?

Don't think that's necessary, especially since the callbacks can now differentiate their behavior based on the opaque user context.

tmm1 commented:
> These still need detailed documentation for their semantics. (E.g. do they block? What values do they return? What happens if a read crosses EOF? Are partial reads allowed? What if a seek goes outside of range? What if no file size is needed? How is it decided whether a stream is for read or write access? Is simultaneous read/write possible (currently it's not in mpv)? Are files on creation with write access truncated or not? etc.)

I've removed `can_write` since I don't have a use-case for it.

I've also split up `mpv_stream_cb_set_callbacks` into distinct functions for each callback, and documented the semantics for each one. This will make it easier to expand the apis in the future (to add back write-support for instance, by creating `mpv_stream_cb_set_write_fn`). The `seek_fn` and `size_fn` are also marked as optional so they can be skipped if the underlying stream does not support those operations.

tmm1 commented:
>> At a loss on how to do this best. stream_t doesn't allow access to mpv_handle or MPContext. I tried putting the callback struct in MPOpts, but it wasn't copied over to the stream_t opts.
>
> It's not necessarily simple - it's not clear how this should be passed through at all. One could argue that this is a good case to add to mpv_global, though I'm not sure if I love that idea.
> 
> In any case, mpv_stream_cb_set_callbacks() should be made thread-safe too.
> 
> MPOpts (and mpv_global as well) is copied for the demuxer thread to avoid thread-safety issues. (The code is in misc.c create_sub_global().)

I couldn't find a better place than `mpv_global`, so that's where the stream_cb_ctx lives now. That's certainly better than a static global in stream_cb.h, and meshes with the purpose of mpv_global as defined in global.h: "The only purpose of this is to make mpv library-safe."

tmm1 commented:
Added a working client_api_example in here which just wraps `fopen`, `fread` and `fclose`.

I think this is looking pretty good, let me know what else needs to happen before this can be merged.

wm4 commented on stream/stream_cb.c:
```diff
@@ -0,0 +1,150 @@
+        if (!p->stream_cb_ctx->size_fn)
+            break;
+        off_t size = p->stream_cb_ctx->size_fn(p->cookie);
+        if (size != (off_t)-1) {
```
The off_t is apparently copy&pasted from stream_file.c, and doesn't make sense here.

wm4 commented on common/global.h:
```diff
@@ -7,6 +7,7 @@
 struct mpv_global {
     struct MPOpts *opts;
     struct mp_log *log;
+    void *stream_cb_ctx;
```
Use an opaque struct type instead of void.

wm4 commented on DOCS/client_api_examples/simple-streamcb/simple-streamcb.c:
```diff
@@ -0,0 +1,93 @@
+
+    mpv_terminate_destroy(ctx);
+    return 0;
+}
```
Hm, I think the sample should be a bit more minimal. Although the original example (which this was copied from) is already pretty minimal, it also demonstrates things that this example doesn't have to. I'd probably remove almost everything but waiting for the shutdown event, and maybe enable the terminal to get log output.

wm4 commented on libmpv/stream_cb.h:
```diff
@@ -0,0 +1,164 @@
+ * mpv_stream_cb_set_*_fn(). Note that your custom callbacks must not
+ * invoke libmpv APIs as that would cause a deadlock.
+ *
+ * Once registered, you can `loadfile cb://myfile`. Your open_fn will be invoked
```
Feels like the first sentence is missing a verb.

wm4 commented on libmpv/stream_cb.h:
```diff
@@ -0,0 +1,164 @@
+ * Opaque context, returned by mpv_get_sub_api(MPV_SUB_API_STREAM_CB).
+ *
+ * This context is global to libmpv, and the callbacks set within it are used
+ * by all mpv_handles created in the same process.
```
Not quite correct. The callbacks are per mpv instance, which usually is a mpv_handle (but there's a way to share the same core across multiple mpv_handles, which is used for e.g. scripting).

wm4 commented on libmpv/stream_cb.h:
```diff
@@ -0,0 +1,164 @@
+ * Open callback used to implement a custom stream.
+ *
+ * @param user_data opaque user data provided via mpv_stream_cb_init()
+ * @param uri name of the stream to be opened
```
This doesn't actually say what it is: it's the remainder part of the cb:// URL. (It's actually described in the doxygen further up, but the doxygen on this function is kind of useless. I have the same complaint about the function of the "cookie".)

tmm1 replied:
Actually the entire URI is passed through now. It's up to the `open_fn` to parse it however it sees fit.

I can flesh out the docs.. wasn't sure how much of the documentation I should repeat here. Is there a way to link to the other docs?

wm4 replied:
> Is there a way to link to the other docs?

Well, we don't "really" use doxygen (it doesn't actually build), so no. But you can just put a hint, something like "see this and there".

wm4 commented on libmpv/stream_cb.h:
```diff
@@ -0,0 +1,164 @@
+
+/**
+ * Read callback used to implement a custom stream.
+ * The semantics of the callback match read(2).
```
IMO this shouldn't reference read(). read() is a POSIX function with very specific semantics, and the details could be easily different from what this function actually does. Worst of all, in POSIX itself read() can behave differently depending on what the FD is, and flags like O_NONBLOCK can change the behavior too.

So you should think of all possible cases (blocking behavior, partial reads, EOF or near-EOF behavior) and describe how it should behave.

On a side note, I think it'd be better if buf was ``void*``, and nbytes ``int64_t``.

wm4 commented on libmpv/stream_cb.h:
```diff
@@ -0,0 +1,164 @@
+
+/**
+ * Seek callback used to implement a custom stream.
+ * The semantics of this callback match lseek(2).
```
This is different from POSIX for certain. The function can return an error if the seek position is beyond the end of the file.

wm4 commented on libmpv/stream_cb.h:
```diff
@@ -0,0 +1,164 @@
+ * @return the resulting offset of the stream
+ * @return -1 if the seek is not allowed or failed
+ */
+typedef int64_t (*mpv_stream_cb_seek_fn)(void *cookie, int64_t offset, int whence);
```
I see no reason for the whence parameter. There's even a separate size callback to remove the need for SEEK_END.

wm4 commented on libmpv/stream_cb.h:
```diff
@@ -0,0 +1,164 @@
+ *
+ * @param cookie opaque cookie identifying the stream,
+ *               returned from mpv_stream_cb_open_fn
+ * @return the total size in bytes of the stream
```
IMO should be allowed to fail (which would mean unknown size).

wm4 commented on stream/stream_cb.c:
```diff
@@ -0,0 +1,150 @@
+static int seek(stream_t *s, int64_t newpos)
+{
+    struct priv *p = s->priv;
+    return (int)p->stream_cb_ctx->seek_fn(p->cookie, (off_t)newpos, SEEK_SET) != -1;
```
Kind of useless casts everywhere. The off_t one is entirely pointless (though maybe harmless, because mpv is always compiled with off_t=64 bits).

wm4 commented on stream/stream_cb.c:
```diff
@@ -0,0 +1,150 @@
+    mpv_stream_cb_read_fn read_fn;
+    mpv_stream_cb_seek_fn seek_fn;
+    mpv_stream_cb_size_fn size_fn;
+    mpv_stream_cb_close_fn close_fn;
```
So, in theory, the API user could call the setters for these fields at any time, even while custom streams are instantiated, so I think there needs to be some synchronization or some concept how to handle this.

Random idea: wrap each setter into a mutex, and also disallow setting these fields while custom streams are instantiated (by keeping a refcount) - this would make a nice API which would make it relatively hard for the API user to hurt himself.

wm4 commented:
I'm considering continuing this. Just with at least 2 changes:
- remove the "sub api" stuff (I've decided that I actually don't like this somewhat pointless indirection, and it should work on the mpv_handle instead)
- making the callbacks non-global - so either have an API that can dynamically create a handle, or making them per-mpv_handle (probably the former)

(It's possible that this conflicts with things I have said before.)

What do you say, do you want to continue, or should I?

tmm1 commented:
If you want to take this over, feel free. Would love to see something like
this get merged.
On Sat, Jun 11, 2016 at 4:33 AM V. Lang <notifications@github.com> wrote:

> I'm considering continuing this. Just with at least 2 changes:
>
>    - remove the "sub api" stuff (I've decided that I actually don't like
>    this somewhat pointless indirection, and it should work on the mpv_handle
>    instead)
>    - making the callbacks non-global - so either have an API that can
>    dynamically create a handle, or making them per-mpv_handle (probably the
>    former)
>
> (It's possible that this conflicts with things I have said before.)
>
> What do you say, do you want to continue, or should I?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mpv-player/mpv/pull/2630#issuecomment-225354800>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAAKB8KZnd9MKkZziYEhGLIW2TU2roRPks5qKpz8gaJpZM4G6Ie2>
> .
>

wm4 commented:
I've adjusted the API: http://sprunge.us/gJMg

Does this sound good to you?

Is it ok if I go through with the change and leave your name as commit author?

tmm1 commented:
> Does this sound good to you?

Looks great.


> Is it ok if I go through with the change and leave your name as commit author?

👍

wm4 commented:
https://git.io/v6seY

tmm1 commented:
Patch looks great. Really like the struct approach.

On Sun, Aug 7, 2016 at 9:12 AM V. Lang <notifications@github.com> wrote:

> https://git.io/v6seY
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mpv-player/mpv/pull/2630#issuecomment-238091192>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKB4hahWS7VIRgg7LOysHU6bqdQPeQks5qdgP1gaJpZM4G6Ie2>
> .
>

wm4 commented:
OK, merged it. I'll push the example part too shorty.

wm4 commented:
The example program (since examples were split out to a separate repo): https://github.com/mpv-player/mpv-examples/commit/ce0491a176bcc29b8655a832061708dde7a38b64

Also, sorry that it took so long!
